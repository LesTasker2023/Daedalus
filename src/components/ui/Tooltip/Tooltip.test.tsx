import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from './index';

describe('Tooltip', () => {
  const renderBasicTooltip = (contentProps = {}) => {
    return render(
      <TooltipProvider>
        <Tooltip open>
          <TooltipTrigger data-testid="tooltip-trigger">Hover me</TooltipTrigger>
          <TooltipContent data-testid="tooltip-content" {...contentProps}>
            Test content
          </TooltipContent>
        </Tooltip>
      </TooltipProvider>
    );
  };

  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = renderBasicTooltip();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches active state snapshot', () => {
      const { container } = renderBasicTooltip({ active: true });
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Basic Functionality', () => {
    it('renders correctly', async () => {
      renderBasicTooltip();
      // Tooltip content is visible when open=true
      expect(screen.getByTestId('tooltip-content')).toBeInTheDocument();
      // Use getAllByText due to Radix duplicating content for accessibility
      const textElements = screen.getAllByText('Test content');
      expect(textElements.length).toBeGreaterThan(0);
    });
  });

  describe('States', () => {
    it('handles active state correctly', () => {
      renderBasicTooltip({ active: true });
      const content = screen.getByTestId('tooltip-content');
      expect(content).toBeInTheDocument();
    });
  });

  describe('Props', () => {
    it('handles sideOffset prop correctly', () => {
      renderBasicTooltip({ sideOffset: 42 });
      const content = screen.getByTestId('tooltip-content');
      expect(content).toBeInTheDocument();
    });
  });

  describe('Accessibility', () => {
    it.skip('can be focused - SKIPPED: Uses Radix focus management', () => {
      expect(true).toBe(true);
    });

    it('has proper ARIA attributes', () => {
      renderBasicTooltip();
      const trigger = screen.getByTestId('tooltip-trigger');
      expect(trigger).toHaveAttribute('aria-describedby');
    });

    it.skip('supports keyboard navigation - SKIPPED: Uses Radix keyboard handling', () => {
      expect(true).toBe(true);
    });

    it('announces changes to screen readers', () => {
      renderBasicTooltip();
      // The role="tooltip" is on the hidden accessibility span
      const tooltipElement = screen.getByRole('tooltip');
      expect(tooltipElement).toBeInTheDocument();
    });

    it('respects reduced motion preferences', () => {
      renderBasicTooltip();
      // Radix handles motion preferences internally
      const content = screen.getByTestId('tooltip-content');
      expect(content).toBeInTheDocument();
    });
  });

  describe('Custom Styling and Props', () => {
    it('accepts custom className', () => {
      renderBasicTooltip({ className: 'custom-class' });
      const content = screen.getByTestId('tooltip-content');
      expect(content).toHaveClass('custom-class');
    });

    it.skip('forwards refs correctly - SKIPPED: Complex ref forwarding', () => {
      expect(true).toBe(true);
    });

    it('spreads additional props', () => {
      renderBasicTooltip({ 'data-custom': 'test-value' });
      const content = screen.getByTestId('tooltip-content');
      expect(content).toHaveAttribute('data-custom', 'test-value');
    });
  });

  describe('Edge Cases', () => {
    it('handles undefined props gracefully', () => {
      renderBasicTooltip({ children: undefined });
      const content = screen.getByTestId('tooltip-content');
      expect(content).toBeInTheDocument();
    });

    it('handles null props gracefully', () => {
      renderBasicTooltip({ children: null });
      const content = screen.getByTestId('tooltip-content');
      expect(content).toBeInTheDocument();
    });

    it('handles empty string props', () => {
      renderBasicTooltip({ className: '' });
      const content = screen.getByTestId('tooltip-content');
      expect(content).toBeInTheDocument();
    });

    it('handles rapid prop changes', () => {
      const { rerender } = renderBasicTooltip({ className: 'class1' });
      rerender(
        <TooltipProvider>
          <Tooltip open>
            <TooltipTrigger>Hover me</TooltipTrigger>
            <TooltipContent data-testid="tooltip-content" className="class2">
              Test content
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
      );
      const content = screen.getByTestId('tooltip-content');
      expect(content).toHaveClass('class2');
    });

    it('handles complex nested content', () => {
      render(
        <TooltipProvider>
          <Tooltip open>
            <TooltipTrigger>Hover me</TooltipTrigger>
            <TooltipContent data-testid="tooltip-content">
              <span>Nested content</span>
              <div>More content</div>
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
      );
      const content = screen.getByTestId('tooltip-content');
      expect(content).toBeInTheDocument();
    });

    it('maintains functionality with many children', () => {
      render(
        <TooltipProvider>
          <Tooltip open>
            <TooltipTrigger>Hover me</TooltipTrigger>
            <TooltipContent data-testid="tooltip-content">
              {Array.from({ length: 100 }, (_, i) => (
                <div key={i}>Item {i}</div>
              ))}
            </TooltipContent>
          </Tooltip>
        </TooltipProvider>
      );
      const content = screen.getByTestId('tooltip-content');
      expect(content).toBeInTheDocument();
    });

    it('handles component unmounting cleanly', () => {
      const { unmount } = renderBasicTooltip();
      expect(() => unmount()).not.toThrow();
    });

    it('preserves functionality after remounting', () => {
      const { unmount } = renderBasicTooltip();
      unmount();
      renderBasicTooltip();
      const content = screen.getByTestId('tooltip-content');
      expect(content).toBeInTheDocument();
    });
  });
});
