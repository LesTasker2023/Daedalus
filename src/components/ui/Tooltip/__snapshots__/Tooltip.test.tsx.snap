// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip > Snapshots > matches active state snapshot 1`] = `
<button
  aria-describedby="radix-«r1»"
  data-state="instant-open"
  data-testid="tooltip-trigger"
>
  Hover me
</button>
`;

exports[`Tooltip > Snapshots > matches default snapshot 1`] = `
<button
  aria-describedby="radix-«r0»"
  data-state="instant-open"
  data-testid="tooltip-trigger"
>
  Hover me
</button>
`;
